rustdoc: Fix invalid suggestions on ambiguous intra doc links

GuillaumeGomez opened the pull request:
Fixes https://github.com/rust-lang/rust/issues/108653.

r? @notriddle

GuillaumeGomez committed:
Fix invalid suggestions on ambiguous intra doc links

GuillaumeGomez committed:
Add regression tests for #108653

jyn514 commented on tests/rustdoc-ui/issue-108653.rs:
```diff
@@ -0,0 +1,35 @@
+// This is ensuring that the UI output for associated items is as expected.
```
Please put both these tests under `intra-doc` with a more descriptive file name.

GuillaumeGomez replied:
I tend to prefer using the issue number which will provide more information rather than a longer file name which will still be lacking.

Moving the files in the sub directory. :+1:

jyn514 replied:
you can include both, `issue-108653-associated-items.rs`

GuillaumeGomez replied:
Good idea, thanks!

notriddle reviewed (changes requested):
Add above test case to test suite, and fix buggy check for item kind.

GuillaumeGomez commented:
I was relying too much on the item on which the intra-doc link was. Big mistake. So instead, I now directly look at the `Res` candidate and start from there directly.

GuillaumeGomez commented:
The problem was that I was stopping the "discovery" when I found one or more match in one location, preventing us to go through all of them. I removed this limitation, fixing this bug and added a new test.

jyn514 commented:
Why is this rewriting so much of the resolution code? Don't we already do the majority of this work elsewhere in the pass?

GuillaumeGomez commented:
> Why is this rewriting so much of the resolution code? Don't we already do the majority of this work elsewhere in the pass?

We could indeed store directly the information we get from the `resolve` method. That could make the code a lot simpler, thanks for the suggestion!

GuillaumeGomez commented:
I'll close this one in favour of https://github.com/rust-lang/rust/pull/109104.
